src/widgets/barchart.rs

```diff
@@ -36,6 +34,9 @@ use super::{Block, Widget};
 /// The chart can have a [`Direction`] (by default the bars are [`Vertical`](Direction::Vertical)).
 /// This is set using [`BarChart::direction`].
 ///
+/// Note: this is the only widget that doesn't implement `Widget` for `&T` because the current
+/// implementation modifies the internal state of self. This will be fixed in the future.
+///
```

Comment:
The barchart is updated when it renders by removing bars that are not visible. If we implemented Widget on @mut Barchart, then the state is changed forever. If the application shrinks the visible area and then grows it (due to screen resize or application layout changes, the removed bars and groups are lost forever.

This can be fixed, but it's not as straightforward as all the other changes which were mostly just repeating the same pattern. The fix is to make the BarChart immutable. An observation on this is that Widgets shouldn't generally mutate parts of their state in destructive ways.

See comments on the implementation of this starting at https://github.com/ratatui-org/ratatui/pull/288#issuecomment-1632466975 for some more context, and also on a followup PR https://github.com/ratatui-org/ratatui/pull/532#pullrequestreview-1643340812 particularly note this comment:

> The remove invisible groups and bars approach (which mutates the state) works because the widget is consumed on render. If we change the widget to be borrowed in the future, then the entire logic for the widget needs to rewritten (this is something that we're considering).

Reply:
Ok makes sense, I should have checked beforehand. :+1: Lets open an issue for that.

Reply:
Issue at https://github.com/ratatui-org/ratatui/issues/870.